zonepb: make subzone DiffWithZone more accurate

otan opened the pull request:
* Subzones may be defined in a different order. We did not take this
  into account which can cause bugs when e.g. ADD REGION adds a subzone
  in the end rather than in the old "expected" location in the subzones
  array. This has been fixed by comparing subzones using an unordered
  map.
* The ApplyZoneConfig we previously did overwrote subzone fields on the
  original subzone array element, meaning that if there was a mismatch
  it would not be reported through validation. This is now fixed by
  applying the expected zone config to *zonepb.NewZoneConfig() instead.
* Added logic to only check for zone config matches subzones from
  active subzone IDs.
* Improve the error messaging when a subzone config is mismatching -
  namely, add index and partitioning information and differentiate
  between missing fields and missing / extraneous zone configs

Resolves #62790

Release note (bug fix): Fixed validation bugs during ALTER TABLE ... SET
LOCALITY / crdb_internal.validate_multi_region_zone_config where
validation errors could occur when the database of a REGIONAL BY ROW
table has a new region added. Also fix a validation bug partition zone
mismatches configs were not caught.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62839)
<!-- Reviewable:end -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX4tqKp9TzXVG-wrDKG:-MX4tqKp9TzXVG-wrDKH:b-sx6qxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d5b5881321ded48e9d2485e1167c541264ccd5/pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs#L323)):*
> ```
> ALTER table regional_by_row CONFIGURE ZONE USING gc.ttlseconds = 10
> 
> statement error zone configuration for table regional_by_row contains incorrectly configured field "num_replicas"
> ```

oops this is wrong, one second


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
> Reviewed 1 of 4 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX4tqKp9TzXVG-wrDKG:-MX7Isol5fjfi72u2B9I:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d5b5881321ded48e9d2485e1167c541264ccd5/pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs#L323)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

oops this is wrong, one second
</blockquote></details>

?
___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs, line 468 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7JXPf8yuEwi0cl5HZ:-MX7JXPf8yuEwi0cl5H_:bhdrj3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs#L468)):*
> ```
> SET override_multi_region_zone_config = false
> 
> statement error attempting to update zone config which is missing an expected zone configuration for partition "ap-southeast-2" of regional_by_row@idx
> ```

Nit: Worth calling `crdb_internal.validate_multi_region_zone_configs()` before this call?
___
*[pkg/config/zonepb/zone.go, line 774 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7LWroCJy1bb7oskUx:-MX7LWroCJy1bb7oskUy:bw6ft1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/config/zonepb/zone.go#L774)):*
> ```Go
> 		if !found {
> 			// Zone configs could be extraneously defined as long as it does not
> 			// differ between fieldLists.
> ```

Nit: I found this comment a bit confusing.  Do we mean that there can be an extra zone config defined so long as it doesn't have any fields in the fieldList set?
___
*[pkg/config/zonepb/zone.go, line 811 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7NCNJ5zYPOvnCKebT:-MX7NCNJ5zYPOvnCKebU:bk98cve) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/config/zonepb/zone.go#L811)):*
> ```Go
> 	}
> 
> 	// Anything remaining in the map can be presumed to be missing.
> ```

Nit: ...provided that there's something set that matches a field in the fieldList?
___
*[pkg/config/zonepb/zone.go, line 826 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7NO2j2HN7KvvEB7yK:-MX7NO2j2HN7KvvEB7yL:b-5afhx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/config/zonepb/zone.go#L826)):*
> ```Go
> 		}
> 	}
> 	return true, DiffWithZoneMismatch{}, nil
> ```

Wow.  Nice catch!  The zone config possibilities are pretty intense.
___
*[pkg/sql/region_util.go, line 456 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7NjxP7qKRGkC_jhoh:-MX7NjxP7qKRGkC_jhoi:bw039i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L456)):*
> ```Go
> func isPlaceholderZoneConfigForMultiRegion(zc zonepb.ZoneConfig) bool {
> 	// Placeholders must have at least 1 subzone.
> 	if len(zc.Subzones) < 1 {
> ```

Nit: Should this be `== 0`, to be a bit more clear?  
___
*[pkg/sql/region_util.go, line 570 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7Ogz-0gC6hu4W_yQd:-MX7Ogz-0gC6hu4W_yQe:b-o8kqa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L570)):*
> ```Go
> 	}
> 
> 	// Mark the NumReplicas as 0 if we have a placeholder.
> ```

Nit: I actually like the old comment better.  If you really prefer the new comment, can you add in a reference to "See zonepb.IsSubzonePlaceholder for why this is necessary."  I personally find this whole placeholder concept tricky, and I suspect others will too.
___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX7PnGDDtkMmEVecfSP:b-hzshjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
> ```Go
> 
> 	// Remove inactive subzones from the comparison.
> 	filteredSubzones := currentZoneConfig.Subzones[:0]
> ```

Nit: Did you add this out of an abundance of caution?  Or did you hit a case where there were inactive indexes with defined zone configs?  I'd think we'd want to root out those cases in addition to handling them here.
___
*[pkg/sql/region_util.go, line 1363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7QeTg2VkhedjR9j1D:-MX7QeTg2VkhedjR9j1E:b-m6nexz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1363)):*
> ```Go
> 	// Mark the expected NumReplicas as 0 if we have a placeholder
> 	// and the current zone config is also a placeholder.
> 	// The latter check is requires as in cases where non-multiregion fields
> ```

Nit: required
___
*[pkg/sql/region_util.go, line 1364 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7QZIW33iZPQpwykVy:-MX7QZIW33iZPQpwykVz:b-e2vo05) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1364)):*
> ```Go
> 	// and the current zone config is also a placeholder.
> 	// The latter check is requires as in cases where non-multiregion fields
> 	// is set on the current zone config, the expected zone config has needs
> ```

Nit: is -> are, also there's an extra "has"
___
*[pkg/sql/region_util.go, line 1367 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7R3AMB5MegNnz13jN:-MX7R3AMB5MegNnz13jO:b7cvopq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1367)):*
> ```Go
> 	// the placeholder marked so that DiffWithZone does not error when
> 	// num_replicas is expectedly different.
> 	if currentZoneConfig.IsSubzonePlaceholder() && isPlaceholderZoneConfigForMultiRegion(expectedZoneConfig) {
> ```

I don't understand this check.  If expected is a placeholder, won't its NumReplicas already be set to 0?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs, line 323 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX4tqKp9TzXVG-wrDKG:-MX8kKV43GHCrO-vCYJ6:bhdfu0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4d5b5881321ded48e9d2485e1167c541264ccd5/pkg/ccl/logictestccl/testdata/logic_test/multi_region_zone_configs#L323)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

?
</blockquote></details>

earlier revision had the wrong result a few lines up.
___
*[pkg/config/zonepb/zone.go, line 774 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7LWroCJy1bb7oskUx:-MX8n0G1-wTu10PoeU5C:b-zi5e3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/config/zonepb/zone.go#L774)):*
>  there can be an extra zone config defined so long as it doesn't have any fields in the fieldList set
much cleaner. comment change.
___
*[pkg/config/zonepb/zone.go, line 826 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7NO2j2HN7KvvEB7yK:-MX8nUnxBcSEPaLnhNxR:b-vch2bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/config/zonepb/zone.go#L826)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Wow.  Nice catch!  The zone config possibilities are pretty intense.
</blockquote></details>

![image.png](https://files.reviewableusercontent.io/images/reviewable/3646147/32oKGDVQ0hF8LWq37pba7ofX/image.png)
___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX8m5pp9_TyAnSXID75:byjh2aa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Did you add this out of an abundance of caution?  Or did you hit a case where there were inactive indexes with defined zone configs?  I'd think we'd want to root out those cases in addition to handling them here.
</blockquote></details>

There were indexes that are being cleaned up during, e.g. transitioning from RBR -> Global which appears to not get cleaned up immediately, in which the zone config hangs around for a while. If these were a concern, I'd say delete this filtering and you'll see a few tests fail, e.g. `go test ./pkg/ccl/telemetryccl -run 'TestTelemetry/multiregion/server'`
___
*[pkg/sql/region_util.go, line 1367 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7R3AMB5MegNnz13jN:-MX8mWJbCuU43_4V-1T6:bwdsnpv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1367)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I don't understand this check.  If expected is a placeholder, won't its NumReplicas already be set to 0?
</blockquote></details>

nope, it is nil. we have to set it by ourselves.
*BUT* if current zone config is not a placeholder because e.g. gc.ttlseconds is set, then we fudge expected zone config to be a placeholder.


<!-- Sent from Reviewable.io -->

otan committed:
zonepb: make subzone DiffWithZone more accurate

* Subzones may be defined in a different order. We did not take this
  into account which can cause bugs when e.g. ADD REGION adds a subzone
  in the end rather than in the old "expected" location in the subzones
  array. This has been fixed by comparing subzones using an unordered
  map.
* The ApplyZoneConfig we previously did overwrote subzone fields on the
  original subzone array element, meaning that if there was a mismatch
  it would not be reported through validation. This is now fixed by
  applying the expected zone config to *zonepb.NewZoneConfig() instead.
* Added logic to only check for zone config matches subzones from
  active subzone IDs.
* Improve the error messaging when a subzone config is mismatching -
  namely, add index and partitioning information and differentiate
  between missing fields and missing / extraneous zone configs

Release note (bug fix): Fixed validation bugs during ALTER TABLE ... SET
LOCALITY / crdb_internal.validate_multi_region_zone_config where
validation errors could occur when the database of a REGIONAL BY ROW
table has a new region added. Also fix a validation bug partition zone
mismatches configs were not caught.

ajstorm reviewed (commented):
> Reviewed 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, and @otan)

___
*[pkg/config/zonepb/zone.go, line 826 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7NO2j2HN7KvvEB7yK:-MX8rnJ01WSgj0T4kaQ0:b-bzrdbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/config/zonepb/zone.go#L826)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

![image.png](https://files.reviewableusercontent.io/images/reviewable/3646147/32oKGDVQ0hF8LWq37pba7ofX/image.png)
</blockquote></details>

The zone configs giveth, and the zone configs taketh away.  Now we just have to figure out how to get them to giveth a bit.
___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX8wuqu4Let2lzwzMx_:b5u3ku5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

There were indexes that are being cleaned up during, e.g. transitioning from RBR -> Global which appears to not get cleaned up immediately, in which the zone config hangs around for a while. If these were a concern, I'd say delete this filtering and you'll see a few tests fail, e.g. `go test ./pkg/ccl/telemetryccl -run 'TestTelemetry/multiregion/server'`
</blockquote></details>

Makes sense that we put this in now, but I'm wondering if there's more work to be done here to clean up those zone configs properly before the schema change, like we do for RBR.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX8xlGR-Kt9vNYgSOSt:b-dqo7be) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Makes sense that we put this in now, but I'm wondering if there's more work to be done here to clean up those zone configs properly before the schema change, like we do for RBR.
</blockquote></details>

We can't do it before, because if the schema change fails we have to "put them back". they need to be live right until the atomic swap of the indexes.
Doing it after makes the most sense atm...


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, and @otan)

___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX97nawBdZSCN6W8vW_:b-saxmkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

We can't do it before, because if the schema change fails we have to "put them back". they need to be live right until the atomic swap of the indexes.
Doing it after makes the most sense atm...
</blockquote></details>

The issue that we hit in the RBR->Global transition originally, was that the ZCs hung around while the indexes were waiting to be GCed.  I _thought_ that we caught those by removing the ZCs when transitioning from RBR.  If you're still hitting them, then something's not right.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX9B9QL-5UbH18UJOzX:b-o5sw31) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

The issue that we hit in the RBR->Global transition originally, was that the ZCs hung around while the indexes were waiting to be GCed.  I _thought_ that we caught those by removing the ZCs when transitioning from RBR.  If you're still hitting them, then something's not right.
</blockquote></details>

Something indeed isn't right (found the bug). I'm inclined to keep this since a) the original bug for that is fixed and b) it doesn't make sense for me to error in this case, and will fix that in a separate issue.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajstorm, and @otan)

___
*[pkg/sql/region_util.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62839#-MX7PnGDDtkMmEVecfSO:-MX9N9tI310DhagljTEM:bjwa7hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f47e5c3db5d262368b5c08fbb5b0187687fe28e6/pkg/sql/region_util.go#L1349)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Something indeed isn't right (found the bug). I'm inclined to keep this since a) the original bug for that is fixed and b) it doesn't make sense for me to error in this case, and will fix that in a separate issue.
</blockquote></details>

:tada: 


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm
